File: workspace/ev_news_collector_php/src/Storage.php
<?php

namespace EVNewsCollector;

class Storage
{
    private \PDO $pdo;

    public function __construct(string $dbPath)
    {
        $this->pdo = new \PDO("sqlite:$dbPath");
        $this->pdo->setAttribute(\PDO::ATTR_ERRMODE, \PDO::ERRMODE_EXCEPTION);
        
        $this->pdo->exec('''
            CREATE TABLE IF NOT EXISTS articles (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                title TEXT,
                link TEXT UNIQUE,
                summary TEXT,
                published TEXT
            )
        ''');
    }

    public function saveToSQLite(array $articles): void
    {
        $count = 0;
        foreach ($articles as $article) {
            try {
                $stmt = $this->pdo->prepare('''
                    INSERT INTO articles (title, link, summary, published)
                    VALUES (?, ?, ?, ?)
                ''');
                $stmt->execute([
                    $article['title'],
                    $article['link'],
                    $article['summary'],
                    $article['published']
                ]);
                $count++;
            } catch (\PDOException $e) {
                // Пропуск дубликатов
                if ($e->getCode() != 23000) { // SQLITE_CONSTRAINT
                    throw $e;
                }
            }
        }
        
        echo "Сохранено $count статей в SQLite.\n";
    }

    public function exportToCSV(array $articles, string $csvPath): void
    {
        $fp = fopen($csvPath, 'w');
        if (!$fp) {
            throw new \Exception("Не удалось открыть файл для записи CSV: $csvPath");
        }
        
        // Заголовки
        fputcsv($fp, ['title', 'link', 'summary', 'published']);
        
        // Данные
        foreach ($articles as $article) {
            fputcsv($fp, $article);
        }
        
        fclose($fp);
        echo "Экспортировано " . count($articles) . " статей в CSV.\n";
    }

    public function exportToJSON(array $articles, string $jsonPath): void
    {
        file_put_contents($jsonPath, json_encode($articles, JSON_UNESCAPED_UNICODE | JSON_PRETTY_PRINT));
        echo "Экспортировано " . count($articles) . " статей в JSON.\n";
    }
}

File: workspace/ev_news_collector/storage.py
import sqlite3
import csv
import json
import logging
from typing import List, Dict

# Настройка логирования
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

def save_to_sqlite(articles: List[Dict], db_path: str):
    try:
        conn = sqlite3.connect(db_path)
        cursor = conn.cursor()
        
        cursor.execute('''
            CREATE TABLE IF NOT EXISTS articles (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                title TEXT,
                link TEXT UNIQUE,
                summary TEXT,
                published TEXT
            )
        ''')
        
        for article in articles:
            try:
                cursor.execute('''
                    INSERT INTO articles (title, link, summary, published)
                    VALUES (?, ?, ?, ?)
                ''', (article['title'], article['link'], article['summary'], article['published']))
            except sqlite3.IntegrityError:
                # Пропуск дубликатов
                pass
                
        conn.commit()
        conn.close()
        logger.info(f"Сохранено {len(articles)} статей в SQLite.")
    except Exception as e:
        logger.error(f"Ошибка при сохранении в SQLite: {e}")

def export_to_csv(articles: List[Dict], csv_path: str):
    try:
        with open(csv_path, 'w', newline='', encoding='utf-8') as csvfile:
            fieldnames = ['title', 'link', 'summary', 'published']
            writer = csv.DictWriter(csvfile, fieldnames=fieldnames)
            writer.writeheader()
            writer.writerows(articles)
        logger.info(f"Экспортировано {len(articles)} статей в CSV.")
    except Exception as e:
        logger.error(f"Ошибка при экспорте в CSV: {e}")

def export_to_json(articles: List[Dict], json_path: str):
    try:
        with open(json_path, 'w', encoding='utf-8') as jsonfile:
            json.dump(articles, jsonfile, ensure_ascii=False, indent=4)
        logger.info(f"Экспортировано {len(articles)} статей в JSON.")
    except Exception as e:
        logger.error(f"Ошибка при экспорте в JSON: {e}")